src/librbd/ExclusiveLock.cc

```diff
@@ -134,11 +134,13 @@ void ExclusiveLock<I>::shut_down(Context *on_shut_down) {
   {
```

Comment:
i modified the code as what you said:
template <typename I>
void ExclusiveLock<I>::shut_down(Context *on_shut_down) {
  ldout(m_image_ctx.cct, 10) << this << " " << __func__ << dendl;

  C_Gather *gather_ctx = new C_Gather(m_image_ctx.cct, on_shut_down);
  {
    Mutex::Locker locker(m_lock);
    assert(!is_shutdown());
    execute_action(ACTION_SHUT_DOWN, gather_ctx);                                                                                                                                      
  }

  // if stalled in request state machine -- abort
  handle_peer_notification(0);

  //allow to delete ExclusiveLock
  gather_ctx->activate();
}

but other issue happened, i get another assertion failure:  
librbd/image/CloseRequest.cc: 143: FAILED assert(m_image_ctx->exclusive_lock == nullptr

this because we delay just only a little, not enough, when we call activate(), the other thread does not yet call ExclusiveLock<I>::handle_shutdown which set m_image_ctx.exclusive_lock to nullptr

so, please give me some ideas.